designs/statefulset-disruption.md

```diff
@@ -0,0 +1,891 @@
+
+Once Karpenter confirms that an instance is terminated, adding this taint to the node object will allow the Attach/Detach Controller to treat the volume as not in use, preventing the 6+ minute delay. 
+
+By modifying Karpenter to apply this taint and wait ~5 seconds until volumes are marked not in use, the following sequence will occur:
+
+![taint_solution](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/d1bc5704-7592-4aa1-a72e-83d33f824b8b)
+
+See [this commit](https://github.com/kubernetes-sigs/karpenter/pull/1294/commits/88134e00ad02863d5a7268bba0b639e21f3f5398) for a proof-of-concept implementation. 
+
+**Pros:**
+
+- Solves 6+ minute delays by default
+- No additional latency before Karpenter's terminator can start instance termination. 
+- Minor latency in deleting Node's finalizer IF karpenter does not treat `shutting down` instance as terminated. (In my tests, a 5 second wait was sufficient for AttachDetach Controller to recognize the out-of-service taint and allow for volume detach)
+- If Kubernetes makes this 6 minute ForceDetach timer infinite by default (As currently planned in Kubernetes v1.32), and the EBS CSI Node Pod is unable to unmount all volumes before improved Node Shutdown Manager times out, the out-of-service taint will be the only way to ensure workload starts on other node 
+
+**Cons:**
+
+- Unavailable until Kubernetes `v1.26`. Customers running Karpenter on Kubernetes ≤ `v1.25` will require solution B1 to be implemented. 
+- Requires Karpenter-Provider-AWS to not treat `Shutting Down` as terminated (Though Karpenter had already planned on this via [PR #5979](https://github.com/aws/karpenter-provider-aws/pull/5979))
+- Problem B's delay still occurs because volumes will not be detached until consolidated instance terminates. 
+
+#### Alternatives Considered
+
+**Customer configuration**
+
+Customers can mitigate 6+ minute delays by configuring their nodes and pod taint tolerations. See [EBS CSI Driver FAQ: Mitigating 6+ minute delays](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#what-steps-can-be-taken-to-mitigate-this-issue) for an updating list of configuration requirements.
+
+A quick overview:
+- Configure Kubelet for Graceful Node Shutdown
+- Enable Karpenter Spot Instance interruption handling
+- Use EBS CSI Driver ≥ `v1.x` in order use the [PreStop Lifecycle Hook](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#what-is-the-prestop-lifecycle-hook)
+- Set `.node.tolerateAllTaints=false` when deploying the EBS CSI Driver
+- Confirm that your stateful workloads do not tolerate Karpenter's `disrupting` taint, and any that any system-critical stateful daemonsets have a lower terminationGracePeriod than the EBS CSI Driver.
+- EBS CSI driver should switch to gRPC liveness probe to ensure kubelet trusts CSI Node Pod during graceful node shutdown. ETA EBS CSI Driver v1.33.0. 
```

Comment:
What's this option. Seems interesting. Is there a link to it? What would this give us?

Reply:
Today, for some reason only with karpenter, even if EBS Node Service container is kept alive due to pre-stop lifecycle hook, liveness probe container is killed, which can lead to Kubelet not relying on EBS node service to unmount (due to failed health check). Jason found out that we can resolve this issue by adding a sleep pre-stop hook to livenessprobe container. Unlike our pre-stop hook, that is a 'dumb' sleep which will delay all node terminations by sleep time. We can avoid having a liveness probe altogether by setting up a gRPC endpoint in our node service that kubelet can health check instead. 

See this section of k8s docs: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe

Thanks to @jmdeal for diving deep and finding sleep workaround, and  Connor knowing about this new feature. 

We want to implement this in next release of driver, which will allow for a real customer workaround. (That doesn't require cx manually editing/patching our daemonset with the sleep prestop for liveness probe)